qutebrowser/components/braveadblock.py

```diff
@@ -0,0 +1,278 @@
+        resource_type: typing.Optional[interceptor.ResourceType] = None,
+    ) -> bool:
+        """Check whether the given request is blocked."""
+        if first_party_url is not None and not first_party_url.isValid():
+            first_party_url = None
+
+        if not first_party_url:
+            return False
+
+        qtutils.ensure_valid(request_url)
+
+        if not config.get("content.blocking.adblock.enabled", url=first_party_url):
+            # Do nothing if adblocking is disabled.
+            return False
+
+        result = self._engine.check_network_urls(
+            request_url.toString(),
+            first_party_url.toString(),
+            resource_type_to_string(resource_type),
+        )
+
+        if not result.matched:
+            return False
+        if result.exception is not None and result.important is None:
+            logger.debug(
+                "Excepting {} from being blocked by {} because of {}".format(
+                    request_url.toDisplayString(), result.filter, result.exception
+                )
+            )
+            return False
+        if _is_whitelisted_url(request_url):
+            logger.debug(
+                "Request to {} is whitelisted, thus not blocked".format(
+                    request_url.toDisplayString()
+                )
+            )
+            return False
+        return True
+
+    def filter_request(self, info: interceptor.Request) -> None:
+        """Block the given request if necessary."""
+        if self._is_blocked(info.request_url, info.first_party_url, info.resource_type):
+            logger.debug(
+                "Request to {} blocked by ad blocker.".format(
+                    info.request_url.toDisplayString()
+                )
+            )
+            info.block()
+
+    def read_cache(self) -> None:
+        if self._cache_path.is_file():
+            logger.debug("Loading cached adblock data: {}".format(self._cache_path))
+            self._engine.deserialize_from_file(str(self._cache_path))
+        else:
+            if (
+                config.val.content.blocking.adblock.lists
+                and not self._has_basedir
+                and config.val.content.blocking.adblock.enabled
+            ):
+                message.info("Run :adblock-update to get adblock lists.")
+
+    def adblock_update(self) -> None:
+        """Update the adblock block lists."""
+        self._done_count = 0
+        self._wip_filter_set = adblock.FilterSet()
+        logger.info("Downloading adblock filter lists...")
+
+        blocklists = config.val.content.blocking.adblock.lists
+        if not blocklists:
+            # Blocklists are None or length zero
+            self._on_lists_downloaded()
+        else:
+            self._finished_registering_downloads = False
+            for i, url in enumerate(blocklists):
+                if i == len(blocklists) - 1:
+                    self._finished_registering_downloads = True
```

Comment:
Reminder to future self: Handle case where the last blocklist entry does not trigger any downloads. This can happen when we have a `file://` URL which points to an empty directory, for example.